README.md
# Moodle Behat Browser configuration utility

## Setup instructions

1. Open your `config.php` in your preferred editor
1. Just before the call to `require(__DIR__ . '/lib/setup.php');` add the following:
```
require_once('/path/to/moodle-browser-config/init.php');
```
1. Initialise behat


## Usage instructions
1. Start up the relevant driver
1. Run behat, specifying one of the defined browsers as a profile, for example:
```
vendor/bin/behat --profile=headlesschrome --config=/Users/nicols/Sites/moodles/w3c/moodledata_behat/behatrun/behat/behat.yml --tags=@javascript
```

## Configuration

Sensible defaults are provided for all drivers. These can be configured in the `config.php` file in the project root
folder. See the included `config-dist.php` for an example configuration.

The following configuration values can be configured:

Driver                | Configuration variable | Description                                                                                      | Default
---                   | ---                    | ---                                                                                              | ---
Selenium              | `seleniumUrl`          | The webdriver URL configured for Selenium                                                        | `http://localhost:4444/wd/hub`
chromedriver          | `chromedriverUrl`      | The webdriver URL configured for chromedriver                                                    | `http://localhost:9515`
Selenium/chromedriver | `chromeBinaryPath`     | The path to the Chrome or Chromium binary. Passed to selenium/chromedriver                       | [Not set]
geckodriver           | `geckodriverUrl`       | The webdriver URL configured for geckodriver                                                     | `http://localhost:4444`
Selenium/geckodriver  | `firefoxBinaryPath`    | The path to the Firefox binary. Passed to selenium/firefoxdriver                                 | [Not set]
edgedriverUrl         | `edgedriverUrl`        | The webdriver URL configured for Edge                                                            | `http://localhost:9515`
Selenium/edgedriver   | `edgeBinaryPath`       | The path to the Edge binary. Passed to selenium/edgedriver                                       | [Not set]
safaridriverUrl       | `safaridriverUrl`      | The webdriver URL configured for Safari                                                          | `http://localhost:4444`
BrowserStack          | `browserstackUrl`      | The webdriver URL configured for Browserstack. The username and password variables are preferred | [Automatically calculated from `browserstackUsername` and `browserstackPassword`]
Browserstack          | `browserstackUsername  | A username used to automatically configure the browserstackUrl                                   | [Not set]
Browserstack          | `browserstackPassword  | A password used to automatically configure the browserstackUrl                                   | [Not set]


An example configuration is available in `config-dist.php`.

## Default profiles

The following are the default profiles included with this utility.

Profile name           | Browser | Driver       | Notes
---                    | ---     | ---          | ---
`chrome`               | Chrome  | Selenium     | chromedriver required
`headlesschrome`       | Chrome  | Selenium     | Similar to `chrome`, but headless
`firefox`              | Firefox | Selenium     | geckodriver required
`headlessfirefox`      | Firefox | Selenium     | Similar to `firefox`, but headless
`chromedriver`         | Chrome  | chromedriver |
`headlesschromedriver` | Chrome  | chromedriver | Similar to `chromedriver`, but headless
`gecko`                | Firefox | geckodriver  |
`headlessgecko`        | Firefox | geckodriver  | Similar to `gecko`, but headless
`edge`                 | MS Edge | Selenium     | Require Selenium 4.0 or later. May require configuration for `edgeBinaryPath`
`headlessedge`         | MS Edge | Selenium     | Similar to `edge`, but headless
`edgedriver            | MS Edge | edgedriver   | Fairly reliable but additional configuration may be required
`headlessedgedriver`   | MS Edge | edgedriver   | Similar to `edgedriver` but headless

### Browserstack Profiles

A small number of Browserstack profiles are included if browserstack credentials have been supplied in the config.php.

Profile name         | Browser | Driver       | Reliability | Notes
---                  | ---     | ---          | ---         | ---
`bs_win_edge`        | MS Edge | BrowserStack | Good        | Similar to `edge`, but running on Browserstack

### Experimental profiles

The following profiles are also included but should be considered experimental. Your mileage may vary.

Profile name         | Browser | Driver       | Reliability | Notes
---                  | ---     | ---          | ---         | ---
`safaridriver`       | Safari  | safaridriver | Very poor   | Click is not supported rendering this driver useless


## Custom profiles

In addition to the Default profiles, you can easily add your own configuration as required.

Create a new `localprofiles.php` file alongside the `init.php` and any configuration you may have.

See the `examples` folder for some examples of these.

```
<?php

use AndrewNicols\Behat\ProfileManager;

return [
    'firefox_nightly' => ProfileManager::getBrowserProfile(
        'firefox',
        ProfileManager::getSeleniumUrl(),
        ProfileManager::supportsW3C(),
        [
            'moz:firefoxOptions' => [
                'binary' => '/path/to/firefox/nightly',
            ],
        ]
    ),
];
```


## Browserstack

It is possible to define and use profiles for BrowserStack. You will need to define the following configuration in your
moodle-browser-config configuration file, for example:

```
$USERNAME = "my_first_tester";
$AUTOMATE_KEY = "Passw0rd!";
$CFG->behat_browserstack_url = "https://$USERNAME:$AUTOMATE_KEY@hub-cloud.browserstack.com/wd/hub";
```

### Custom Browserstack profiles

Use the W3C Protocol Generator for BrowserStack: https://www.browserstack.com/automate/capabilities?tag=selenium-4

Browserstack has options in the `bstack:options` capability, but some capabilities may need to be specified at the top level.

This can be achieved via passing a `capabilities` entry to the array used in `getBrowserProfile()`:

```
<?php

use AndrewNicols\Behat\ProfileManager;

return [
    'bs_osx_firefox' => ProfileManager::getBrowserProfile(
        'firefox',
        ProfileManager::getBrowserStackUrl(),
        ProfileManager::supportsW3C(),
        [
            'bstack:options' => [
                'os' => 'Windows',
                'osVersion' => '10',
                'local' => true,
            ],
            'capabilities' => [
                'browser' => 'Firefox',
                'browserVersion' => 'latest-beta',
            ],
        ]
    ),
];
```

Further examples can be seen in the examples folder.


Changelog.md
# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/).

## [Unreleased]
### Fixed
- The MicrosoftEdge browser can now be started directly, via Selenium, or via BrowserStack


init.php
<?php
// This file is part of Moodle - http://moodle.org/
//
// Moodle is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// Moodle is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.
//
// You should have received a copy of the GNU General Public License
// along with Moodle.  If not, see <http://www.gnu.org/licenses/>.

/**
 * Behat Browser configuration.
 *
 * @package    test
 * @copyright  2020 Andrew Nicols <andrew@nicols.co.uk>
 * @license    http://www.gnu.org/copyleft/gpl.html GNU GPL v3 or later
 */

use AndrewNicols\Behat\ProfileManager;

require_once(__DIR__ . '/src/ProfileManager.php');

if (!ProfileManager::isBehatCliUsage()) {
    // This is not a behat call. Do not add the profile manager.
    return;
}

$browserSetup = new ProfileManager();
$browserSetup->addStandardProfiles();
$browserSetup->addCustomProfiles();


examples/browserstack.php
<?php

use AndrewNicols\Behat\ProfileManager;

return [
    'bs_win10_firefox' => ProfileManager::getBrowserProfile(
        'firefox',
        ProfileManager::getBrowserStackUrl(),
        ProfileManager::supportsW3C(),
        [
            'bstack:options' => [
                'os' => 'Windows',
                'osVersion' => '10',
                'local' => true,
            ],
            'capabilities' => [
                'browserName' => 'Firefox',
                'browserVersion' => 'latest-beta',
            ],
        ]
    ),
    'bs_osx_firefox' => ProfileManager::getBrowserProfile(
        'firefox',
        ProfileManager::getBrowserStackUrl(),
        ProfileManager::supportsW3C(),
        [
            'bstack:options' => [
                'os' => 'OS X',
                'osVersion' => 'Big Sur',
                'local' => true,
            ],
            'capabilities' => [
                'browserName' => 'Firefox',
                'browserVersion' => 'latest-beta',
            ],
        ]
    ),
];


config-dist.php
<?php
// This file is part of Moodle - http://moodle.org/
//
// Moodle is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// Moodle is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.
//
// You should have received a copy of the GNU General Public License
// along with Moodle.  If not, see <http://www.gnu.org/licenses/>.

/**
 * Behat Browser configuration.
 *
 * @package    test
 * @copyright  2020 Andrew Nicols <andrew@nicols.co.uk>
 * @license    http://www.gnu.org/copyleft/gpl.html GNU GPL v3 or later
 */

return (object) [

    //
    // Configuration for Selenium.
    //
    // A default value for the selenium URL:
    //  'seleniumUrl' => 'http://localhost:4444/wd/hub',

    //
    // Configuration for chromedriver.
    // Chromedriver is the driver used for Google Chrome.
    //
    // A default value for the chromedriver URL:
    //  'chromedriverUrl' => 'http://localhost:9515',
    //
    // You can also specify a binary to use.
    //  'chromeBinaryPath' => '',

    //
    // Configuration for geckodriver.
    // Geckodriver is the driver used for Mozilla Firefox.
    //
    // A default value for the geckodriver URL:
    //  'geckodriverUrl' => 'http://localhost:4444',
    //
    // You can also specify a binary to use.
    //  'firefoxBinaryPath' => '',

    //
    // Configuration for safaridriver.
    // Safaridriver is the driver used for Apple's Safari Browser.
    //
    // A default value for the safaridriver URL:
    //  'safaridriverUrl' => 'http://localhost:4444',

    //
    // Configuration for edgedriver.
    // Edgedriver is the driver used for the non-legacy MS Edge.
    //
    // A default value for the edgedriver URL:
    //  'edgedriverUrl' => 'http://localhost:9515',
    //
    // You can also specify a binary to use for Microsoft Edge.
    //  'edgeBinaryPath' => '',

    //
    // Browserstack Configuration
    //
    // You can provide a username and password for browserstack.
    // These will be used to generate the correct browserstack URL.
    //  'browserstackUsername' => 'myusername';
    //  'browserstackPassword' => 'myautomationkey';
    //
    // Alternatively you can specify a full Browserstack Webdriver URL.
    //  'browserstackUrl' => "https://USERNAME:KEY@hub-cloud.browserstack.com/wd/hub";
    //
    // If both the username/password, and the url are provided, then only the url is used.
];
